### PR TITLE
PCHR-2883: Improve Absence Types calculation unit selector

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

This PR disables the Absence Type calculation unit selector if it cannot be changed (for example, when such an absence type has already been used in leave requests).

## Technical Details

In PHP form builder the `disabled` attribute is now conditionally set:

```php
$this->addSelect(
  'calculation_unit',
  $this->getCalculationUnitSelectorParams(),
  TRUE
);
// ...
private function getCalculationUnitSelectorParams () {
  $calculationUnitSelectorParams = [
    'options' => $this->getCalculationUnitOptions(),
    // ...
  ];

  if ($this->absenceTypeHasEverBeenUsed()) {
    $calculationUnitSelectorParams['disabled'] = 'disabled';
  }

  return $calculationUnitSelectorParams;
}
```

However, this is not enough since the disabled selector would not provide a value on form submit which makes such an Absence Type updating impossible due to the failing backend validation. 

Removing `disabled` attribute from the `<select>` would not work since the CRM selector watches the original `<select>` and unsets `disabled` attribute for itself as well. Same happens if you then apply `disabled` attribute to the CRM selector - the original `<select>` would set `disabled` attribute to itself.

So another method needs to be performed on the front-end: adding `<input type hidden>` just before the `<select>`, watch `<select>` value and set it to the hidden input.

```js
function supportValueSubmissionOfAllSelectorsIfDisabled () {
  var $selectors = $('select.crm-select2');

  $.each($selectors, function (index, select) {
    supportValueSubmissionOfSelectorIfDisabled($(select));
  });
}

function supportValueSubmissionOfSelectorIfDisabled ($select) {
  var $inputHidden = $('<input type="hidden" name="' +
    $select.attr('name') + '" value="' +
    $select.val() + '" />');

  $select.before($inputHidden);
  $select.removeAttr('name');
  $select.on('change', function () {
    $inputHidden.val($select.val());
  });
}

$(document).ready(function() {
  // ...
  supportValueSubmissionOfAllSelectorsIfDisabled();
});
```

## Comments

Changing a selector to a label instead would not work since flickering will happen.

-----

✅ Manual Tests - passed
✅ PHP Unit Tests - passed
⏹ CSS distributive files - not needed
⏹ Jasmine Tests - n/a
⏹ JS distributive files - not needed
⏹ Backstop Tests - omitted

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/templates/CRM/HRLeaveAndAbsences/Form/AbsenceType.tpl [hunk 1]
```diff
@@ -320,14 +320,28 @@
...
+                }
+
                 $(document).ready(function() {
                   initToilControls();
                   initCarryForwardControls();
                   initColorPicker();
                   initDeleteButton();
-                  initCalculationUnitControls()
+                  initCalculationUnitControls();
+                  supportDisabledSelector($('select[name=calculation_unit]'));
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/templates/CRM/HRLeaveAndAbsences/Form/AbsenceType.tpl]: Something like `supportValueSubmissionOfDisabledSelector()` would be more appropriate.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Form/AbsenceType.php [hunk 2]
```diff
@@ -275,6 +271,25 @@ private function getDAOFieldAttributes($field) {
...
+   * Returns parameters needed for the calculation unit selector
+   *
+   * @return array
+   */
+  private function getCalculationUnitSelectorParams () {
+    $calculationUnitSelectorParams = [
+      'options' => $this->getCalculationUnitOptions(),
+      'option_url' => NULL,
+      'label' => ts('Calculate Leave in')
+    ];
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Form/AbsenceType.php]: The function `getCalculationUnitOptions()` is no longer needed, the reason I created the function was to find a workaround to not disable the calculation unit field and make sure that the selected option is the only option.

We can simply change the the `$calculationUnitSelectorParams` array to

> $calculationUnitSelectorParams= [
>       'option_url' => NULL,
 >      'label' => ts('Calculate Leave in')
 >    ];

Civi will automatically build the select options from the calculation unit option group.
Also we should remove the `getCalculationUnitOptions()` function.

**Comment by igorpavlov** [comment]
@deb1990 I amended the implementation slightly. The new implementation supports enabling selectors back (if needed) and fixes all selectors as well for the future.

File: uk.co.compucorp.civicrm.hrleaveandabsences/templates/CRM/HRLeaveAndAbsences/Form/AbsenceType.tpl (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/templates/CRM/HRLeaveAndAbsences/Form/AbsenceType.tpl]: Done, but slightly differently.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Form/AbsenceType.php (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Form/AbsenceType.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/templates/CRM/HRLeaveAndAbsences/Form/AbsenceType.tpl [hunk 3]
```diff
@@ -320,14 +320,44 @@
...
+                  });
+                }
+
+                /**
+                 * Adds <input type="hidden"> before disabled <select>
+                 * with the same, watches <select> value and sets to the <input>
+                 *
+                 * @param {jQueryElement} $select - the selector to be amended
+                 */
+                function supportValueSubmissionOfSelectorIfDisabled ($select) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/templates/CRM/HRLeaveAndAbsences/Form/AbsenceType.tpl]: All the function definitions should be after the init function(`document.ready` in this case). Please move the document.ready function to the beginning.

**Pull request merged**